Concluding python exercise pull to master

Yairzeh opened the pull request:
pull!!!!!!!!!!!!!!!!!!!!!!!!

Yairzeh committed:
Trying merge requests

0-:

Yairzeh committed:
Delete ConcludingExcersize.zip

Yairzeh committed:
Update and rename ConcludingExcersize.py to ConcludingExercise.py

liambas commented:
היי שלום,
קודם כל עבודה ממש ממש מגניבה - התחלה יפה.
כמובן שיש עוד הרבה מאוד עבודה - אל תתבאס אם זה ייקח עוד כמה ימים, זו ממש לא הגרסא הסופית שאני מכוון אלייה.
כמה דברים בסיסיים ב-cr/pr - זאת הדרך שלי לוודא אותך ולתת הערות - ברמת השורה. כלומר, יהיו לי הערות כלליות בהתחלה ואז מלא הערות לכל פונקציה וכו.
תקרא על זה קצת - זה יחסית נוח.
דבר אחרון, בגדול הקונבנציה זה לכתוב באנגלית, וזה גם יותר נוח - לכן מעכשיו כל מה שאכתוב יהיה באנגלית
ודבר ממש ממש אחרון - ייקח לי זמן לעבור על זה (רוצה להגיד שעה) - תן לי לסיים את ההערות ואז תתחיל לעבוד על זה, שלא נתנגש
בהצלחה.

liambas commented:
general notes and comments

- [x] Pleae add a new option [0] or [3] to create new account.
- [ ] In general, the Team's convention is to use Numpy documentation style - its a widely used style, read about it and how pycharm can help you create it. For now don't change it but you should get used to it.
- [x] I made a lot of comments that are general conventions, change what i said everywhere, even if i only mentioned it on one function.
- [x] fle names should be of the format 'file_name', and not 'FileName'

liambas committed:
.

liambas commented on ConfigurationFile.json.py:
```diff
@@ -1 +1,17 @@
-
+[
+    {
+        "account_number": "1234",
+        "password": "abcd",
+        "balance": 2000
+    },
+    {
+        "account_number": "5678",
+        "password": "1234",
+        "balance": -200
+    },
+    {
+        "account_number": "4321",
+        "password": "yair",
+        "balance": 400
+    }
+]
```
Json is a good file format for this task. One small thing - Everything in Json should be a dict - that means even the "outer" list - so wrap it in a dictiinary.

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+2: similarly to how all the different actions are done, I made a "save & exit" action in the action menu.
+ Additionally I added an "Exit out of program" command to the main menu, so you could turn off the program entirely.
+'''
+FILE_DIRECTORY = 'ConfigurationFile.json'
```
This isnt a "file directory" its the config file. The convention is to name the configuration file as config.json and then the variable will be something like CONFIG_FILE_PATH

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+
+# Functions:
+
+def main_menu():
+    """
+    shows the main menu and its options, asks for input on what to do and returns it.
+    """
+    what_to_do = input('\nWelcome! Here are your options:\n\n'
+                       '[1] Log-in\n'
+                       '[2] Exit out of program\n'
+                       'Enter your choice: ')
+    return what_to_do
```
1. Function names should always be verbs - i like them in present simple (read_file, load_data, run_algorithm, parse_data) and so on and so on. main_menu is not a good name, and it isnt the meaning of the function. instead it should be something like "get_user_starting_input or something like that
2. the variable what_to_do is wrong for several reasons - first, the name is not indicative. it isnt the naming convention. Instead it should be something like "user_input_field" or something like that. But more importantly, you don't need this variable - you declare it and then return it right after. Instead, just return the input() - its shorter and nicer

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+    return what_to_do
+
+
+def open_file():
+    """
+    opens the JSON configuration file and returns it.
+    """
+    with open(FILE_DIRECTORY, 'r') as file:
+        return json.load(file)
```
1. Nice function - its a generally good util function. In general, you shouldn't be afraid to use multiple files. Create a new file utils.py and dump all the utilities in there, like this function and other utils if you have them.
2.  this is not a "pure function" - it uses variables from the outside (FILE_DIRECTORY). instead, if its called open_file() then its best to pass file_path as a variable.
3. rename to load_json_file

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+    dumps the config file, with indent=2
+    """
+    with open(FILE_DIRECTORY, 'w') as file:
+        json.dump(config, file, indent=2)
```
this is also a util function - can you only dump accounts? what if i want to dump some other stuff? this should be in utils.py.
also - try to generalize, its always a good practice - for example, if i wrote save_to_config ill make it save anything i want, not just "accounts"

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+    """
+    returns the account balance!
+    """
+    return account['balance']
```
You decided to use a dict to represent 'account'. This is limited and you have to pass the account and its just ugly. 
I want you to start thinking about OOP - object oriented programing. Think of logical things that can be their own "type" - for example an Account class is a classic OOP example - i want you to implement Account as a class and have all these methods inside the class (deposit,withdraw,dump_to_json,show_balance and so on and so on

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+                   '4. Change password\n'
+                   '5. Save & Exit\n'
+                   'Enter your choice: ')
+    return choice
```
same here you dont need choice variable, just return the input, and change the function name

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+
+        else:
+            print('\nInvalid input!')
+            continue
```
This is a classic problem - you have many many if elses that depend on some user input and your'e stuck with endless if else statments. What you should do is read about a "method factory" and implement it (think of a dict that stores methods). change it.

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+        if correct - returns the account.
+        if incorrect - prints "Wrong password!" and raises ValueError.
+    """
+    success = False
+    account_number = input('\nEnter account number: ')
+    for account in accounts:
+        if account_number == account['account_number']:
+            account_index = accounts.index(account)
+            success = True
+            continue
+    if not success:
+        print('\nAccount does not exist!')
+        raise ValueError('Account does not exist!')
+    else:
+        password_attempt = input('\nEnter password: ')
+        if password_attempt == accounts[account_index]['password']:
+            return accounts[account_index]
+        print('\nWrong password!')
+        raise ValueError('Wrong password!')
```
1. This logic is very complicated. you dont need the success variable, just check it some account matches and if not raise the value error (and if it does exist then go to the password attempt.

2. using account index is very arabic - it leads to errors (what if somehow the accounts get mixed and arent sorted? or many other bugs that can occur). When using Account as a class, you can implement check credentials which makes this function deprecated

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+    else - changes the password.
+    """
+    new_password = input('Enter your new password: ')
+    if new_password == account['password']:
+        print('Password is the same!')
+    else:
+        account['password'] = new_password
```
print('Password is the same!') if new_password == account['password'] else account['password'] = new_password

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+
+# Program:
+
+start = True
+while start:
```
just do while True, its acceptable

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+        print('Invalid input!')
+        continue
+
+    '''
+    opens the JSON file, and uses the open account to ask for account number and password,
+    if all was correct, go on with code.
+    if something was wrong - continues (re-enters the main menu)
+    -
+    if went on with code (user number and password were correct) - sets action_start to True 
+    and begins a new while loop for the action menu.
+    '''
```
It's uncommon to put documentation in the middle of the function. It should be at the start before the code

liambas commented on ConcludingExercise.py:
```diff
@@ -1 +1,201 @@
+    try:
+        deposit_amount = float(input('\nEnter amount to deposit: '))
+        account['balance'] += deposit_amount
+        print('\nMoney Withdrawn Successfully!')
```
Your'e depositing not withdrawing

Yairzeh committed:
Update and rename ConfigurationFile.json.py to config.py

- Changed name from ConfigurationFile.json to config,json.
- Updated config so everything is a dictionary and not a list.

Yairzeh committed:
Update and rename ConcludingExercise.py to ATM_Program.py

changed basically everything according to the comments.
main large changes:
- Added an Account class and methods,
- Changes choice menus to use method factories.
- Changed variable function and method names to be clearer.
- Moved some utility functions to a separate file called utils.py

Yairzeh committed:
Add utils.py seperate from the main file.

- Includes function to load a .json file,
- Includes function to dump any data into a .json file.

Yairzeh committed:
Update and rename ATM_Program.py to atm_program.py

- Removed capital letters from file name.
- Added a method in the startup menu to create a new account.

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+CONFIG_FILE_PATH = 'config.json'
+
+
+# Creating an Account class and methods:
```
Comments are usually good, but sometimes they're just annoying and don't add anything - like this comment (everyone sees you created an account class, no need to say it). Side note - chatgpt likes these types of comments very much

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+# Importing relevant functions from utils.py
```
also here, useless comment

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+# Importing relevant functions from utils.py
+from utils import load_json_file
+from utils import dump_to_json
```
can just do from utils import load_json_file,dump_to_json

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+        withdraws the account from account, unless:
+        - withdrawal amount is negative, greater than balance or not a number, then raises a ValueError correspondingly.
+        """
+        try:
+            withdrawal_amount = float(input('\nEnter Withdrawal amount: '))
+            if withdrawal_amount > self.__balance:
+                raise ValueError('Not enough balance!')
+            if withdrawal_amount < 0:
+                raise ValueError('Withdrawal amount must be positive!')
+            self.__balance -= withdrawal_amount
+        except ValueError:
+            print('\nInput must be a number!')
```
I think the first thing you should do is try: withdrawl amount... and check for value error. then after you made sure its actually a float you can do the 2 if statments (so try except FIRST then the 2 if statments outside) - its clearer and more organized

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+        deposits the account from account, unless:
+        - deposit amount is negative or not a number, then raises a ValueError correspondingly.
+        """
+        try:
+            deposit_amount = float(input('\nEnter deposit amount: '))
+            if deposit_amount < 0:
+                raise ValueError('Deposit amount must be positive!')
+            self.__balance += deposit_amount
+        except ValueError:
+            print('\nInput must be a number!')
```
same here, do the try except first, then outside do the if statment

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+        """
+        saves the account to the config file using the dump_to_json from utils.py.
+        """
+        account_data = {
+            'password': self.__password,
+            'balance': self.__balance,
+        }
+        accounts = load_json_file(file_path)
+        accounts[self.account_number] = account_data
+        dump_to_json(file_path, accounts)
```
I like the idea of having a save method on an account - but its funny to me that you have to load the file, update the data then dump the file again. Maybe extract these 3 lines of loading, entering and dumping to a util function?

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+            raise
+
+
+# Method Factories:
```
again useless comments

liambas commented:
- [x] Move Account class to a new file account.py because its a main class and deserves it's own file
- [x] remove all the uselss comments you have - if they don't add anytning then theyre annoting and litter the code
- [x] I need you to start thinking in OOP - object oriented programming. Everything in python is supposed to be classes (in reality we don't always use classes, we mostly use functions, but still we aim for things that can be a class, to be a class)
Because of this, i think you should add another class (just an idea i had, you need to test it) - UserAccounts or AtmAccoutnts that has a list of all the accounts (save the account Classes) and then inside that class you can have stuff like create new account (just append an Account to the list) and a "log in method" and also a dump to json

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+            return cls(account_number, account_password, account['balance'])
+        except ValueError as error:
+            print(error)
+            raise
```
From a logical point of view, this function makes no sense. you have an Account class that has a method called "load account" (which is a good method) but it takes the file with every account and checks it and does parsing and all sorts of things that shouldnt be there. instead, look at my other comment about the UserAccounts class, which will solve this problem easily

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+    }
+
+
+def get_action_method(account, file_path):
+    """
+    Method factory for the action menu choices.
+    """
+    return {
+        '1': account.print_balance,
+        '2': account.withdraw,
+        '3': account.deposit,
+        '4': account.change_password,
+        '5': lambda: account.save_account_to_config(file_path)
+    }
```
perhaps its better to have this method factory inside the account class, im not sure though, what do you think

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+                action(cls, file_path)
+            else:
+                print('\nInvalid input, please try again!')
+        except ValueError as error:
+            continue
```
you just continue?

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+
+
+# Use the program with Account as the class and CONFIG_FILE_PATH as config directory.
+start_program_loop(Account, CONFIG_FILE_PATH)
```
read about if __name__=='main' in python and implement it here

liambas commented on config.py:
```diff
@@ -0,0 +1,14 @@
+{
+  "1234": {
+    "password": "abcd",
+    "balance": 2000
+  },
+  "5678": {
+    "password": "1234",
+    "balance": -200
+  },
+  "4321": {
+    "password": "yair",
+    "balance": 400
+  }
+}
```
why is this config.py and not config.json?

liambas commented on utils.py:
```diff
@@ -0,0 +1,17 @@
+    dumps the config file, with indent=2
+    """
+    with open(file_path, 'w') as file:
+        json.dump(data, file, indent=2)
```
put indent into the function arguments: dump_to_json(file_path, data,indent=2):

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,224 @@
+    account = cls.load_account(file_path)
+    action_methods = get_action_method(account, file_path)
+
+    while True:
+        choice = get_user_action()
+        action = action_methods.get(choice)
+        if action:
+            try:
+                action()
+            except ValueError as error:
+                print(error)
+                raise
+            if choice == '5':
+                break
+        else:
+            print('\nInvalid input, please try again!')
```
here there are some things to change around. in get user action, you should include the action itself and check if its valid. What i mean is, the first if else (if action else invalid input.. should be in a similar option. think of it like this:
1. get the function the user wanted (user action) - this also checkts that the user input is valid
2. run the action he chose
3. check that everything is alright and repeat

Yairzeh committed:
upload of account,py and useraccounts,py

- account.py: python file containing the Account class
- useraccounts.py: python file containing the new UserAccounts class

Yairzeh committed:
Update atm_program.py

Yairzeh committed:
Rename config.py to config.json

maybe I accidentally changed it to .py somehow? not really sure

Yairzeh committed:
Update utils.py

liambas commented on config.json:
```diff
@@ -0,0 +1,14 @@
+    "password": "yair",
+    "balance": 400
+  }
+}
```
im now thinking that "config" isnt a good name for this. usually config files are for general system configurations (constants, things you might want to change and so on). maybe here the correct name would be accounts.json or accounts_ifno.json

liambas commented on utils.py:
```diff
@@ -0,0 +1,26 @@
+
+def load_and_dump_to_json(file_path, key, value, indent=2):
+    """
+    loads the config file, and dumps the key:value pair into it.
```
This is a good function. 2 things:

1. this should be called inser_value into json file. i think this is a better name but you decide
2.  change the documentation - this doesnt load a "config" file -t it loads ant json

liambas commented on account.py:
```diff
@@ -0,0 +1,85 @@
+        self.__password = password
+        self.__balance = balance
+
+    def __eq__(self, other):
```
nice

liambas commented on account.py:
```diff
@@ -0,0 +1,85 @@
+        """
+        dictionary containing the methods performed by the main program's action menu
+        """
+        from useraccounts import UserAccounts
```
?

liambas commented on account.py:
```diff
@@ -0,0 +1,85 @@
+        if deposit_amount < 0:
+            print('Deposit amount must be positive!')
+            raise ValueError
+        self.__balance += deposit_amount
```
Heres something to think about - these 2 functions are almost the same - they perform the same operations, one is -= and one is += but everything else apart from the "deposit" or "withdraw"  messages is the same. Try combining them. The idea is to remove repeated code.

liambas commented on account.py:
```diff
@@ -0,0 +1,85 @@
+        try:
+            deposit_amount = float(input('\nEnter deposit amount: '))
+        except ValueError:
+            print('Input must be a number!')
```
can you just raise ValueError('Input must be a number!') instead of printing?

liambas replied:
I now see that it crashes the program. do you catch the value error?

liambas commented on account.py:
```diff
@@ -0,0 +1,85 @@
+        }
+        return self.account_number, account_values
+
+    def get_action_method(self, user_accounts, file_path):
```
Youre passing file path but now using it. This should also be called something else - maybe get user actions? Also this should be a class property - read about it

liambas commented on account.py:
```diff
@@ -0,0 +1,85 @@
+        return hash(self.account_number)
+
+    def get_password(self):
+        return self.__password
```
This is a classic thing - you declar password as "protected" (__...) but then you make a function that just returns it. so why protect it? its ok to do it but then you dont need this function, just use account.password

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+from utils import load_json_file, load_and_dump_to_json
```
This file name should be user_accounts. the rule is:
1. for file names, always use underscore: file_name my_file and so on
2. for functions use the same notation: my_cool_function
3. for classes use this notation: MyCoolClass, UserAccounts

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+from account import Account
+
+
+class UserAccounts:
+    def __init__(self, file_path):
+        """
+        defines an accounts list containing each account in the config as an Account class.
+        """
+        self.file_path = file_path
+        self.__accounts = [Account(account_number, data['password'], data['balance'])
+                           for account_number, data in load_json_file(file_path).items()]
```
Ok, you implemented this well. 
what if i have a list of accounts instead of a file? for that reason i think the cooler way to implement this, which i want you to try is the following:
the class accepts a list of accounts and stores them instead of accpting the file. If you have a  file (which you do), implement a "class method" (read about it !!!) that loads the file and returns the class. This is a common thing to do when using classes in python and i find it really clean and nice.

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+        dumps all accounts into the config file.
+        """
+        for account in self.__accounts:
+            load_and_dump_to_json(self.file_path, account.get_account_data()[0], account.get_account_data()[1])
```
This is ugly, i dont like how u use get_account_data, why not just pass in the password and balance? Also, instead of doing it for each account cant you do it all at one? just create a big dict and dump it

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+        and asks for password, adds it to the accounts list and dumps the list into the config file.
+        """
+        account_number = input("\nEnter new account number: ")
+        if Account(account_number, None, None) in self.__accounts:
```
This check is super ugly!!! the useraccounts class should have a list of all the user accounts (easy to create) and just check if the account number is in the list

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+        account_password = input("\nEnter new account password: ")
+        account = Account(account_number, account_password)
+        self.__accounts.append(account)
+        self.dump_accounts_to_config()
```
You dump all the accounts to config even though you just added one account, this is really not efficient

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+        correct returns the account as an Account class.
+        """
+        account_number = input("\nEnter account number: ")
+        if Account(account_number, None, None) not in self.__accounts:
```
again same thing here, ugly. just add a new function in the class and call it check if accounts exists or something. DONT repeat code.

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+        if Account(account_number, None, None) not in self.__accounts:
+            print('Account does not exist!')
+            raise ValueError
+        account = self.__accounts[self.__accounts.index(Account(account_number, None, None))]
```
why not just do self.__accounts[index where the account number matches]

liambas commented on useraccounts.py:
```diff
@@ -0,0 +1,49 @@
+        self.__accounts.append(account)
+        self.dump_accounts_to_config()
+
+    def load_account(self):
```
this isnt really loading anything. its either log_into_account or get_account_by_number

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,87 @@
+            break
+
+
+def exit_program():
```
This is a util function, should be in utils

liambas commented on atm_program.py:
```diff
@@ -0,0 +1,87 @@
+                   '[3] Create a new account\n'
+                   'Enter your choice: ')
+    if choice not in ['1', '2', '3']:
+        print('Invalid choice!')
```
You should just raise valueerror(invalid choice), you do this all around the code, fix it

Yairzeh committed:
Update account.py

Yairzeh committed:
Update atm_program.py

Yairzeh committed:
Rename config.json to accounts_info.json

Yairzeh committed:
Update and rename useraccounts.py to user_accounts.py

Yairzeh committed:
Update utils.py
